### PR TITLE
draft post about str += being quadratic

### PR DESCRIPTION
**Pull request by cfbolz**

### CODE REVIEW

File: posts/2023/01/string-concatenation-is-quadratic.rst [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+character copies.
+
+Repeated string concatenations are in principle also quadratic in CPython, but
+CPython has an optimization_ that makes them sometimes not quadratic, which is
+what makes this program not too slow in CPython.
+
+.. _optimization: https://docs.python.org/2/whatsnew/2.4.html#optimizations
+
+In order to fix the problem on PyPy it's best to use a list for the string
+parts, which as the right amortized O(1) complexity for ``.append`` calls, and
```
Comment (mattip) [inline posts/2023/01/string-concatenation-is-quadratic.rst:65]:
```suggestion
parts, which has the right amortized O(1) complexity for ``.append`` calls, and
```

**Review by mattip — COMMENTED** [review]
Looks good, thanks. We should be able to emit a warning if we see this pattern.

Do you want to mention the [blurb in the pypy docs](https://doc.pypy.org/en/latest/cpython_differences.html?highlight=join#performance-differences) about this?

**Comment by cfbolz** [comment]
heh, I actually forgot that it's also in the FAQ. will link that.

**Comment by cfbolz** [comment]
not FAQ, CPy differences. anyway, linked, and fixed the typo. will post soon.

**Pull request merged**